flang/lib/Lower/OpenMP.cpp

```diff
@@ -87,19 +88,32 @@ genOMP(Fortran::lower::AbstractConverter &absConv,
     auto &firOpBuilder = absConv.getFirOpBuilder();
     auto currentLocation = absConv.getCurrentLocation();
     auto insertPt = firOpBuilder.saveInsertionPoint();
+
+    // Clauses.
+    mlir::Value numThreads;
+
+    const auto &parallelOpClauseList =
+        std::get<Fortran::parser::OmpClauseList>(blockDirective.t);
+    for (const auto &clause : parallelOpClauseList.v) {
+      if (const auto &numThreadsClause =
+              std::get_if<Fortran::parser::OmpClause::NumThreads>(&clause.u))
+        // OMPIRBuilder accepts `NUM_THREAD` clause as a `Value`.
+        numThreads = absConv.genExprValue(
+            *Fortran::semantics::GetExpr(numThreadsClause->v));
+    }
     llvm::ArrayRef<mlir::Type> argTy;
-    mlir::ValueRange range;
-    llvm::SmallVector<int32_t, 6> operandSegmentSizes(6 /*Size=*/,
-                                                      0 /*Value=*/);
-    // create and insert the operation.
+    // FIXME: Add support for remaining clauses.
+    mlir::ValueRange range(numThreads);
+    // Create and insert the operation.
     auto parallelOp = firOpBuilder.create<mlir::omp::ParallelOp>(
         currentLocation, argTy, range);
-    parallelOp.setAttr(mlir::omp::ParallelOp::getOperandSegmentSizeAttr(),
-                       firOpBuilder.getI32VectorAttr(operandSegmentSizes));
-    parallelOp.getRegion().push_back(new Block{});
+    parallelOp.setAttr(
+        mlir::omp::ParallelOp::getOperandSegmentSizeAttr(),
+        firOpBuilder.getI32VectorAttr({0, (numThreads ? 1 : 0), 0, 0, 0, 0}));
```

Comment:
```suggestion
        firOpBuilder.getI32VectorAttr({0, (hasNumThreads ? 1 : 0), 0, 0, 0, 0}));
```

I think you just need to track if the `num_threads` clause is present. If it's not just just put a 0 here and it should be fine. This code snippet is with the previous suggestion as well. This should work just fine.

Reply:
The problem is not here, it's at line 106  `mlir::ValueRange range(numThreads)`. If NUM_Threads clause is not present, this expression triggers assertion: Since range is **NOT** as empty but with **NULL TYPE** which is not equivalent to empty.
Basically you can construct a null Value as
```
mlir::Value someValue;
//but, constructing ValueRange with NULL as
mlir::ValueRange(someValue); // Not allowed triggers assertion.
```
```
mlir/include/mlir/IR/Value.h:81: 
bool mlir::Value::isa() const [with U = mlir::BlockArgument]: Assertion `*this && "isa<> used on a null type."'
 failed.
```

Reply:
So you should not create this range if `num_threads` is not there. As I understood from river's comment in discord is that a null value in a ValueRange is considered as a value so if you have no value you should not have the range. Basically you create the op the same way as before when there was no clause lowered. 

So if num_threads is not here you have: 
```mlir
mlir::ValueRange range;
```
and if you have a clue you just override range with a new object including numThread value.

Reply:
By the way, I'm going to work on the same things on the OpenACC side so if I see something that could help I'll let you know.